### ISSUE TITLE
Agent doesn't connect to server

### ISSUE THREAD

**Issue by FilGov** [Ylianst/MeshCentral#5345]
**Describe the bug**
MeshCentral agent doesn't connect to MeshCentral server.

**To Reproduce**
1. Install MeshCentral server on server.
2. Install MeshCentral client on remote device.
3. Remote device never shows up on MeshCentral server.

**Expected behavior**
The remote device shows up on MeshCentral server.

**Screenshots**
![MeshCentral](https://github.com/Ylianst/MeshCentral/assets/49760773/87d0f704-c31b-48db-9a1f-8505c5e26b14)

**Server Software (please complete the following information):**
 - OS: Windows Server 2016
 - Virtualization: none
 - Network: LAN
 - Version: 1.1.1.1
 - Node: 14.17.1

**Client Device (please complete the following information):**
 - Device: Desktop
 - OS: Windows 11 Pro
 - Network: Remote over WAN
 - Browser: Google Chrome
 - MeshCentralRouter Version: N/A

**Remote Device (please complete the following information):**
 - Device: Desktop
 - OS: Windows 10 22H2
 - Network: Local to Meshcentral
 - Current Core Version (if known): ?

**Additional context**
None.

**Your config.json file**
```
{
   "settings": {
      "lanonly": true, 
      "_minify": true, 
      "selfupdate": true, 
      "clickonce": true, 
      "webrtc": true
   }, 
   "domains": {
      "": {
         "newaccounts": false
      }
   }
}
```

**Comment by FilGov** [comment]
It showed up after ~1 hour. Any idea why it took so long and how to speed it up?

**Comment by FilGov** [comment]
Yes. 

How/Where to set `cert`? We removed `lanonly`, restarted MeshCentral, re-downloaded the agent, and tried again, issue persists.

**Comment by si458** [comment]
> Yes.
> 
> How/Where to set `cert`? We removed `lanonly`, restarted MeshCentral, re-downloaded the agent, and tried again, issue persists.

follow the sample basic config - https://github.com/Ylianst/MeshCentral/blob/master/sample-config.json

anything with an `_` is ignore, so in `_cert` cause, rename it to `cert`

**Comment by FilGov** [comment]
Done, issue persists.

**Comment by SomeGuru** [comment]
Clarify to me your config for connection from the agent side... Are you saying the device is going through a VPN? If so then your administration on the VPN could be scanning/logging extensive details before releasing the agent to the wild, and if this is the case you might have to work with your IT department. Correct me if I am wrong that your config states a WAN VPN.

-SomeGuru

**Comment by FilGov** [comment]
Yes, the device "where the MeshCentral server web interface is being accessed" is going through a VPN to connect to the LAN where the MeshCentral server and clients are. No, the device "where the MeshCentral agent was installed" isn't going through a VPN. It's in the same LAN as the MeshCentral server. No agent is going through any VPN. I'm the IT department.

**Comment by silversword411** [comment]
So the possible issues you have outside of the meshcentral server are:

1. VPN routing problems
2. DNS resolution thru VPN
3. Your mesh server is using self-signed certs (bad ssl errors in browser)
4. What you blocked out in your browser address bar is very short, so I'm suspecting non-public DNS names
5. Traffic routing between agent and server
6. Validating a simple and direct mesh agent to mesh server connection actually works.

You haven't provided any network topology, routing information, ICMP testing results, DNS resolution validation, TLS information or any technical data to show you know how to diagnose your routing infrastructure and instead just say "issue persists"

As a long time veteran, I can only conclude this is an implementation problem, not a meshcentral server issue or bug.

I'm happy to consult and diagnose your problem...by the hour. Feel free to open a ticket for support: https://h.ineedhelpwithtech.com/

**Comment by FilGov** [comment]
1. There's no VPN routing involved in the agent communication, the MeshCentral server and agent are in the same LAN.
2. See #1.
3. Correct.
4. You're suspecting wrong, it's a private IP address.
5. See #1. The agent device can ping the server device directly (no hops/routers involved).
6. The MeshCentral agent to MeshCentral server connection is simple and direct (see #5).

I provided everything in the template and everything I was requested. You apparently read none of it and instead just list possible issues outside MeshCentral (many of which invalid), suggest I don't know how to diagnose my routing infrastructure, and assume "this is an implementation problem".

As a long-time veteran, you should read before you post. I already wrote most of this. As a CompTIA and Microsoft-certified administrator, I conclude this is a MeshCentral issue or bug, not an implementation problem. As a long-time veteran, I conclude you don't know what the issue or bug is, so you just blame the user and the implementation.

Not interested in your services, thank you.

**Comment by FilGov** [comment]
1. There's no VPN routing between agent and server, they're in the same LAN.
2. See 1.
3. Correct.
4. You're suspecting wrong, it's a private IP address.
5. There's no routing between agent and server, they're in the same LAN.
6. The MeshCentral agent to MeshCentral server connection is simple and direct (see 1 and 5).

I have provided everything requested in the template and replies. The network topology is simple, just switches between agent and server. There's no routing involved. The agent device can ping the server device. DNS resolves fine and the issue persists when `cert` is set to the server IP address. We're using the default self-signed certificate. I don't just say "issue persists", I describe the issue and provide everything requested in the template and replies. I say "issue persists" because it does.

Your conclusion is unhelpful and contradicts the data.

Not interested in your services, thank you.

**Comment by si458** [comment]
**@FilGov please dont be rude to our community members, we are only trying to help!**

please can you share your `config.json` as of now,
if the `lanonly` is set as true, then you need to make sure the remote computer can ping the meshcentral server but also the meshcentral server can receive the multicast messages the agent sends out to be discovered,

so the network needs to be layer 2 based (in mikrotik EOIP, or cisco equilvent), doing a vpn with lanonly will cause issues, so this needs to be `false` and then you set the `cert` value to a dns name that both the server and the remote agent can resolve to the same IP and also ping that IP too

you also say your using windows server as the server, have you tried turning off the firewall completely?

**Comment by si458** [comment]
@FilGov i can see why we have all got confused! i will need to change the bug report template
your laptop to access to the meshcentral web panel is **Remote over WAN (VPN)**
and the remote agent and meshcentral are **Local to Meshcentral (so LAN)**

you say `(some agents connect)` so do you have multiple machine already connected to meshcentral and they connected straight away?
is it always a certain machine that wont connect to meshcentral or takes forever to show up?
how did you setup the meshcentral server? the installer exe or npm install?
a fluke test (try disabling the firewall, restart meshcentral, try uninstalling the agent, then reinstalling the agent?)

**Comment by FilGov** [comment]
@si458 I agree the bug report is confusing (that's why I defined client and remote device) and should be changed. I'm using a desktop, but that's otherwise correct.

Yes, I have multiple machines already connected to MeshCentral. No, they didn't connect straight away.
No, it isn't always a certain machine that won't connect to MeshCentral or takes forever to show up. Multiple machines won't connect to MeshCentral and the ones that do take "forever" to show up.
I set up the MeshCentral server with the installer executable.
I disabled the firewall, restarted MeshCentral, uninstalled the agent(s) (there were 2 entries in the Control Panel ("Mesh Agent" and "MeshCentral Agent - Remote Control Software") but the latter became invalid after uninstalling the former), then reinstalled the agent - it immediately connected. 
I then re-enabled the firewall, restarted MeshCentral, uninstalled the agent(s) from another machine, then reinstalled the agent - it didn't show up. I then accessed the web panel from inside our network from a machine - it still didn't show up. I then re-disabled the firewall and restarted MeshCentral - it immediately connected.
I then installed the agent on another machine via PsExec without uninstalling anything - it immediately connected.
I then unsuccessfully tried reproducing this on another machine. I noticed the installer didn't detect previous installations of "Mesh Agent" despite there being one, so I re-installed it - it detected and updated it and connected.

So there are apparently 2 issues: the server setup missing firewall rule(s) (which one(s)?) and the client installation missing previous installations.

**Comment by si458** [comment]
@FilGov I'm not 100% sure as I only run linux?
I always believed it was port 80 (http), 443 (https), and 4433 (amt)
I'm not sure who made the windows installer?
I will look into it when I get chance for you!
But I'm guessing for the time being, keep the firewall disabled, 
Or
Play around allowing different stuff through the firewall and see what u can get working

**Comment by tomsik-radek** [comment]
I have mine in Docker behind a reverse proxy (nginx), in WAN mode, so my solution might not apply, but I solved a similar issue with certURL parameter. Without that agents wouldn't connect and Assistant would get stack on "authenticating"

```
"domains": {
    "": {
      [redacted]
      "certUrl": "https://meshcentral.tld.com:443",
      [redacted]
```

**Comment by FilGov** [comment]
> @FilGov I'm not 100% sure as I only run linux? I always believed it was port 80 (http), 443 (https), and 4433 (amt) I'm not sure who made the windows installer? I will look into it when I get chance for you! But I'm guessing for the time being, keep the firewall disabled, Or Play around allowing different stuff through the firewall and see what u can get working

TCP ports 80, 443, 4433, 16990, and 16991 and UDP ports 16990 and 16991 are already open. 
We unfortunately can't keep the firewall disabled, it's a production server.
I unsuccessfully created a rule for C:\Program Files\Open Source\MeshCentral\winservice\daemon\meshcentral.exe.

**Comment by si458** [comment]
@FilGov i think you will need to create a firewall rule for `node.exe`
as the `meshcentral.exe` is just the daemon exe to RUN `node.exe`

**Comment by FilGov** [comment]
> I have mine in Docker behind a reverse proxy (nginx), in WAN mode, so my solution might not apply, but I solved a similar issue with certURL parameter. Without that agents wouldn't connect and Assistant would get stack on "authenticating"
> 
> ```
> "domains": {
>     "": {
>       [redacted]
>       "certUrl": "https://meshcentral.tld.com:443",
>       [redacted]
> ```

Tried setting `certURL` to the FQDN, DNS name, and IP address with and without the port number in hybrid and WAN mode, both installer and server always crash. Reset to LAN mode, the installer still crashes but the server doesn't.

**Comment by FilGov** [comment]
> @FilGov i think you will need to create a firewall rule for `node.exe` as the `meshcentral.exe` is just the daemon exe to RUN `node.exe`

Where's `node.exe`? Please advise, I cannot find it. Thank you.

**Comment by FilGov** [comment]
> @FilGov i think you will need to create a firewall rule for `node.exe` as the `meshcentral.exe` is just the daemon exe to RUN `node.exe`

Done, issue persists.

**Comment by tomsik-radek** [comment]
> > I have mine in Docker behind a reverse proxy (nginx), in WAN mode, so my solution might not apply, but I solved a similar issue with certURL parameter. Without that agents wouldn't connect and Assistant would get stack on "authenticating"
> > ```
> > "domains": {
> >     "": {
> >       [redacted]
> >       "certUrl": "https://meshcentral.tld.com:443",
> >       [redacted]
> > ```
> 
> Tried setting `certURL` to the FQDN, DNS name, and IP address with and without the port number in hybrid and WAN mode, both installer and server always crash. Reset to LAN mode, the installer still crashes but the server doesn't.

If you try just "connect" with Agent instead of installing, what do you see in "Connection details"? Are your Mesh server IP and port listed properly?

**Comment by FilGov** [comment]
This: 

![Connection Details](https://github.com/Ylianst/MeshCentral/assets/49760773/f2bf059d-dcfc-4a9d-a591-aa86244db114)

No, our Mesh server IP address and port aren't listed properly.

**Comment by tomsik-radek** [comment]
> This:
> 
> ![Connection Details](https://user-images.githubusercontent.com/49760773/268668211-f2bf059d-dcfc-4a9d-a591-aa86244db114.png)
> 
> No, our MeshCentral server IP address and port aren't listed properly.

Is yours in LAN mode? Try disabling both forced LAN and WAN modes, ie put it to Hybrid

**Comment by FilGov** [comment]
Can't say, the server crashes almost immediately. Tried replacing "meshcentral.tld.com" with the server IP address, FQDN, and DNS name, same thing.

**Comment by tomsik-radek** [comment]
> Can't say, the server crashes with that config.json. Tried replacing "meshcentral.tld.com" with the server IP address, FQDN, and DNS name, same thing.

Crashes with what error message? Unable to parse config.json?

**Comment by FilGov** [comment]
None. No, the service just stops.

**Comment by SomeGuru** [comment]
What information do you have in the localDiscovery?

    "localDiscovery": {
      "name": "Test.domain.tld",
      "info": "Listed entry aligns to internal xyz department and is a managed remote management service for the company."
    },

The tld is based on your specific domains information. This should help your DNS server identify the asset on the network. Also what is the traceroute look like from your device to the server, and what does the DNS server response take to lookup the server on the network?

SomeGuru

**Comment by FilGov** [comment]
None. We have no such information in our config.json. We added it and changed "Test.domain.tld" to our domain name, issue persists. 

Traceroute: 1     1 ms    <1 ms    <1 ms  <FQDN> [<private IP address>]

DNS server response:

Reply from <private IP address>: bytes=32 time<1ms TTL=128
Reply from <private IP address>: bytes=32 time=1ms TTL=128
Reply from <private IP address>: bytes=32 time=1ms TTL=128
Reply from <private IP address>: bytes=32 time=1ms TTL=128

**Comment by tomsik-radek** [comment]
> None. No, the service just stops.

We will need some logs, we can't go off of that. I assume you are running it as a Windows service? Try running that same config.json with the NPM MeshCentral version.

Also, nothing in the Event log?

**Comment by tomsik-radek** [comment]
A big facepalm for me, sorry. I read your initial post and you are running in LAN mode. In that case no certs, no TLDs should be necessary. Unfortunately I can't help you wit that more as I have never tried LAN mode from myself, I have no use for it.

You could always run WAN mode but just not expose it to the outside world.

**Comment by FilGov** [comment]
> > None. No, the service just stops.
> 
> We will need some logs, we can't go off of that. I assume you are running it as a Windows service? Try running that same config.json with the NPM MeshCentral version.
> 
> Also, nothing in the Event log?

You assume correctly. What's the NPM MeshCentral version? 

The event log lists the following events:

Event ID 1000 - "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js stopped running."
Event ID 1000 - "Restarted 3000 msecs after unexpected exit; attempts = 1"
Event ID 1000 - "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js stopped running."
Event ID 1000 - "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js stopped running."
Event ID 0 - "Child process [4244 - C:\Program Files\nodejs\node.exe  --harmony "C:\Program Files\Open Source\MeshCentral\node_modules\node-windows\lib\wrapper.js" --file "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js" --log "MeshCentral wrapper" --grow 0.5 --wait 2 --maxrestarts 3 --abortonerror n --stopparentfirst undefined] finished with 1"

**Comment by FilGov** [comment]
> A big facepalm for me, sorry. I read your initial post and you are running in LAN mode. In that case no certs, no TLDs should be necessary. Unfortunately I can't help you wit that more as I have never tried LAN mode from myself, I have no use for it.
> 
> You could always run WAN mode but just not expose it to the outside world.

No problem, I understand the VPN detail is confusing, thanks anyway.

The issue persist in WAN mode.

**Comment by tomsik-radek** [comment]
> > > None. No, the service just stops.
> > 
> > 
> > We will need some logs, we can't go off of that. I assume you are running it as a Windows service? Try running that same config.json with the NPM MeshCentral version.
> > Also, nothing in the Event log?
> 
> You assume correctly. What's the NPM MeshCentral version?
> 
> The event log lists the following events:
> 
> Event ID 1000 - "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js stopped running."
> 
> Event ID 1000 - "Restarted 3000 msecs after unexpected exit; attempts = 1"
> 
> Event ID 1000 - "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js stopped running."
> 
> Event ID 1000 - "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js stopped running."
> 
> Event ID 0 - "Child process [4244 - C:\Program Files\nodejs\node.exe --harmony "C:\Program Files\Open Source\MeshCentral\node_modules\node-windows\lib\wrapper.js" --file "C:\Program Files\Open Source\MeshCentral\winservice\winservice.js" --log "MeshCentral wrapper" --grow 0.5 --wait 2 --maxrestarts 3 --abortonerror n --stopparentfirst undefined] finished with 1"

Is that the result of the XML view on the crash event? Sadly this doesn't tell us much.  "--log Meshcentral Wrapper", I wonder if that spits a log file somewhere.

This is the NPM version. https://www.youtube.com/watch?v=GsQbWZmRRAU Try running your config there, this might all be an issue with NodeJS. 

MC is an NPM application. The Windows installer registers the Node app as a Windows service (and similar thing happens with systemD services on Linux). I run the unofficial, but community supported, Docker variant. https://hub.docker.com/r/typhonragewind/meshcentral

I have actually never considered your setup. MeshCentral locally in a LAN, and devices are visible in it when on a VPN. Honestly, I would still do it the WAN mode, and instead of outside FQDN I would point it at a local IP. So Agents would be talking to https://meshcentral.domain.local.

**Comment by FilGov** [comment]
No, this is the result of the general view of the crash event after trying

```
  "settings": {
    "cert": "meshcentral.tld.com",
    "WANonly": true,
    [etc]
  },
  "domains": {
    "": {
      "certUrl": "https://meshcentral.tld.com:443",
      [etc]
    }
  }
}
```

and replacing "meshcentral.tld.com" with the server IP address, FQDN, and DNS name. This is unrelated to the issue itself (agents not connecting). Nothing crashes when the agents don't connect. They simply don't connect.

I tried doing it in WAN mode and pointing it at a local IP, all nodes go offline and the issue persists.

**Comment by tomsik-radek** [comment]
> No, this is the result of the general view of the crash event after trying
> 
> ```
>   "settings": {
>     "cert": "meshcentral.tld.com",
>     "WANonly": true,
>     [etc]
>   },
>   "domains": {
>     "": {
>       "certUrl": "https://meshcentral.tld.com:443",
>       [etc]
>     }
>   }
> }
> ```
> 
> and replacing "meshcentral.tld.com" with the server IP address, FQDN, and DNS name. This is unrelated to the issue itself (agents not connecting). Nothing crashes when the agents don't connect. They simply don't connect.
> 
> I tried doing it in WAN mode and pointing it at a local IP address, the issue persists and all nodes go offline.

The thing is, when I was setting up my server, if I had these cert settings incorrect I had a docker log full of messages about "communication rejected". We need those logs you are getting as well. Try to set it up in the "debug" mode I linked, see what error messages you get there.

**Comment by FilGov** [comment]
That concerns WAN mode. We're using LAN mode. You said it yourself: "in that case no certs, no TLDs should be necessary".

**Comment by tomsik-radek** [comment]
Well, kinda. My point is that without the proper error messages/logs we can't really move anywhere because we don't know what's wrong. Run it in NPM mode so you have a console, then we can do something with it.

**Comment by FilGov** [comment]
I'll try to do that when I can. Meanwhile I'll just temporarily disable the firewall, it's easier and quicker. Maybe meanwhile someone discovers what firewall rule is missing, since that's apparently the culprit.

**Comment by tomsik-radek** [comment]
Well by default they should just run on 443. Are you sure your server is on a private network? Your config is empty so no custom ports there

**Comment by FilGov** [comment]
TCP port 443 is already open. Yes, I am sure our server is on a private network. Correct, no custom ports in our config or anywhere else, we changed no ports anywhere.

**Comment by tomsik-radek** [comment]
Then we really need that console output or error logs, because something is failing

**Comment by SomeGuru** [comment]
Can you run via administrator command prompt with the --debug command and share details from that to us?

SomeGuru

**Comment by FilGov** [comment]
Run what? Please clarify, as --debug (apparently) doesn't work with the MeshCentral agent installer. Thank you.

**Comment by tomsik-radek** [comment]
> Run what? Please clarify, as --debug (apparently) doesn't work with the MeshCentral agent installer. Thank you.

I assume Guru means run the server with - debug

**Comment by FilGov** [comment]
Our server doesn't support that, it's a Windows service. I'll try to run it via command prompt when I can. Meanwhile I'll just temporarily disable the firewall. It's easier and quicker. Maybe meanwhile someone finds out what firewall rule is missing, since that's apparently the culprit.

**Comment by tomsik-radek** [comment]
> Our server doesn't support that, it's a Windows service. I'll try to run it via command prompt when I can. Meanwhile I'll just temporarily disable the firewall, it's easier and quicker. Maybe meanwhile someone discovers what firewall rule is missing, since that's apparently the culprit.

That service has to launch an NPM version of MeshCentral somewhere. Where I don't know as I don't use it, but service isn't magic, it is just an automatic exe

**Comment by FilGov** [comment]
I also don't know where as I didn't develop it. I just know our server runs as a Windows service and doesn't support it. I tried running the .exe the service points to with `-debug` from the command line but it failed, it doesn't support the command. I'll just continue temporarily disabling the firewall when installing agents.
